dependency types

mathstuf opened the pull request:
---
So here are my current notes for where things need to be changed to facilitate different types of dependencies. By no means exhaustive since this is a breadth-first search from `depends_on` to one or two levels for what needs modification/inspection. The API seems to make lines easy to draw (`pkg.dependencies` will need to be a method, but there aren't *too* many uses of that).

Looking through, it looks as though the best storage is `Dict[name, (spec, reason)]` rather than `Dict[reason, Dict[name, spec]]` (as I started in the one commit when playing around with the `depends_on` code).

Are there any other dep types that would matter other than build-time and run-time (run-time for a library being a dep from netcdf onto hdf5 due to a link dependency)? The only bits I saw were "we need it to build" and "we need it to use it" dependencies.

@tgamblin

alalazo commented:
I'll just leave my two cents for the analysis part of this feature :

- when building a package, the run-time DAG for that package is a subgraph of the build-time DAG
- when building a package, build dependencies matter only for the root of the run-time DAG

Then : should build dependencies be part of the computed hash? On the one hand I think we should permit to upgrade build tools seamlessly, i.e. without spack complaining that installed packages depends on the older version (and for that I would say 'no'). On the other hand it would be nice to keep track  in a weaker way of the tools used in the build for the sake of reproducibility. Are there already some thoughts on how to handle this?

eschnett commented:
The build dependencies should be logged in the installed package's `.spack/spec.yaml`.

mathstuf commented:
> when building a package, the run-time DAG for that package is a subgraph of the build-time DAG

Not necessarily. NumPy might be a runtime-used bit, but not needed to build and install.

> when building a package, build dependencies matter only for the root of the run-time DAG

Right. Package A needing Python to build does not mean that B (which depends on A) needs Python available as well.

> should build dependencies be part of the computed hash?

This, I think, is tied to the SDK/platform question raised elsewhere. If those build dependencies can alter the ABI, yes, otherwise, no. The only situations I can think of where this would happen are contrived, so I'd lean towards "no" unless someone can come up with a use case where the ABI would be affected (other than toolchain builddeps).

alalazo commented:
> > when building a package, the run-time DAG for that package is a subgraph of the build-time DAG

> Not necessarily. NumPy might be a runtime-used bit, but not needed to build and install.

@mathstuf : hmm, I am not sure I got your point. Let's say I package `py-foo` which is a pure python package that depends on `numpy` at run-time. Then I'll say that `py-foo`:
```python
extends(python)
depends_on(numpy)
```
From my perspective `numpy` is also needed to build and install `py-foo` (at least within spack) as it takes part in computing the hash for the unique installation prefix. Am I missing something?

mathstuf commented:
Ah, true, it would need to be there for hash computation. I was thinking of ParaView which calls NumPy from C++ code and does not need it during the build, but it does need to advertize which NumPy is expected (though it only cares that it is compatible, not a specific version).

tgamblin commented:
@mathstuf: I'll add some notes to the spots you marked, but I wanted to get an answer out here before the conversation drops out of my inbox (sigh).

> Are there any other dep types that would matter other than build-time and run-time (run-time for a library being a dep from netcdf onto hdf5 due to a link dependency)? 

It's worth considering this from the perspective of what Spack would do differently for different dependency types.  Spack could do these things for deps:

1. Add `-L` flags and `RPATHs` in the compiler wrapper, and add to `LD_LIBRARY_PATH` at run time.
2. Add `-I` flags in he compiler wrapper.
3. Add to `PATH` at build time.
4. Add to `PATH` at run time (or, more generally, load the module for the package - this might include other paths and variables that are needed, like `QTDIR`)
5. Include in `spec` DAG passed to build (and in `spec.yaml`)
6. Include in version hash.

Currently Spack does all of those for all deps.  As a first cut I might break that down like so:
* **Build dependencies**: 3, 5
* **Link dependencies**: 1, 2, 3, 5, 6
* **Run dependencies**: 4, 5, 6

>> should build dependencies be part of the computed hash?
> This, I think, is tied to the SDK/platform question raised elsewhere. If those build dependencies can alter the ABI, yes, otherwise, no.

I think we're on the same page here, and this is exactly why compilers are treated specially in Spack.  They're build deps that directly affect the ABI.  Spack doesn't (yet) do any ABI checking when it builds, but it does try to keep the compilers consistent by looking up in the DAG. We also allow different compilers for different packages in the same build and we've thought about better ABI checking.

Build deps are similar in that they can be different on different nodes in the DAG.  Some packages might need Python but only to run a build script (PETSc).  Another package might link with PETSc, but it shouldn't have to use an entirely different version of Python just because PETSc ran its build system using another.

This will have interesting ramifications for concretization.  Currently when you load up a dependency spec from a `spec.yaml`, you get ALL of its dependencies.  You probably want to remove the build deps from the dependency's DAG and only force transitive link and run deps to be included in the new root's DAG, along with the root's build deps.  That way you can avoid the weird PETSc issue above.  If you include build deps in the version hash, I think you have to force consistency of build deps across the DAG.  That, or you have to start including multiple specs with the same name in a single spec DAG, but that complicates things like `spec['python']` --- which Python should it return?  I'd rather not get into that because it seems like needless extra complexity.

So, my take is that the build deps for package A should be documented in `A/.spack/spec.yaml`, but they shouldn't be included in `dag_hash()` or included automatically as transitive dependencies.

I haven't fully thought through run dependencies, but an example would be that `curl` is a run dependency of `git`, and decent versions of `tar` would be a run dependency of `spack`.  We could talk about whether run deps should be hashed or not... but if you don't hash them then we would need a better mechanism to load the right run dependencies in modules.  Spack doesn't currently have a way to determine compatibility in those cases.

FWIW, [Maven](https://maven.apache.org) has a decent taxonomy of dependency types that is more extensive than the one above.  It calls them [dependency scope](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope).  Note that I don't advocate adopting Maven's XML syntax, but the model is decent and I like the term "scope" for this.

Maven's `provided` scope is an interesting one, as it corresponds to what people are doing when they explicitly leave a dep out of `package.py` to pull in the system version of it.  Personally, I think we don't need this, and it can be taken care of by the externals support in #120 and #309.  In other words, I think `provided` is a property of the dependency and not of the relationship between the dependent and the dependency.

That said, to make externals more usable, it would be nice to generalize detection of external build and run dependencies from the current way compilers are detected.  #309 adds support for multiple compiler detection strategies -- perhaps we could add system package detection strategies in a similar way.

Thoughts?

mathstuf commented:
On Wed, Jan 20, 2016 at 12:08:11 -0800, Todd Gamblin wrote:
> @mathstuf: I'll add some notes to the spots you marked, but I wanted
> to get an answer out here before the conversation drops out of my
> inbox (sigh).

I know that story :) .

> > Are there any other dep types that would matter other than
> > build-time and run-time (run-time for a library being a dep from
> > netcdf onto hdf5 due to a link dependency)? 
> 
> It's worth considering this from the perspective of what Spack would
> do differently for different dependency types.  Spack could do these
> things for deps:
> 
> 1. Add `-L` flags and `RPATHs` in the compiler wrapper, and add to `LD_LIBRARY_PATH` at run time.

Any reason for rpath *and* `LD_LIBRARY_PATH`?

> 2. Add `-I` flags in he compiler wrapper.
> 3. Add to `PATH` at build time.
> 4. Add to `PATH` at run time (or, more generally, load the module for
>    the package - this might include other paths and variables that are
>    needed, like `QTDIR`)

It only does this for directories that exist right? For example, don't
add libjpeg to PATH if `$jpeg_root/bin` doesn't exist? A separate branch
in any case.

> 5. Include in `spec` DAG passed to build (and in `spec.yaml`)
> 6. Include in version hash.
> 
> Currently Spack does all of those for all deps.  As a first cut I
> might break that down like so:
> 
> * **Build dependencies**: 3, 5

Build deps should probably lose #5 (they'll pick up that step if they're
also link deps).

> * **Link dependencies**: 1, 2, 3, 5, 6
> * **Run dependencies**: 4, 5, 6

The only use case which differentiates link and run dependencies are
"plugins" which is handled by the `extends` directive already, so it
does seem that build and run are sufficient (build dep for linking
usually turns into a link dependency and is the common case, so that is
the default…where I've now added a comment for the rationale).

> >> should build dependencies be part of the computed hash?
> > This, I think, is tied to the SDK/platform question raised
> > elsewhere. If those build dependencies can alter the ABI, yes,
> > otherwise, no.
> 
> I think we're on the same page here, and this is exactly why compilers
> are treated specially in Spack.  They're build deps that directly
> affect the ABI.  Spack doesn't (yet) do any ABI checking when it
> builds, but it does try to keep the compilers consistent by looking up
> in the DAG. We also allow different compilers for different packages
> in the same build and we've thought about better ABI checking.

The only tools which I can think which do this:

  - compilers;
  - standard libraries (usually bundled with the compiler, but things
    like STLSoft exist);
  - specialized compilers (e.g., CUDA); and
  - wrapping tools (e.g., swig, shiboken).

Everything else is, if it's only a build dependency, irrelevant to the
ABI.

> Build deps are similar in that they can be different on different
> nodes in the DAG.  Some packages might need Python but only to run a
> build script (PETSc).  Another package might link with PETSc, but it
> shouldn't have to use an entirely different version of Python just
> because PETSc ran its build system using another.

Correct. My thought was to only look up runtime deps when generating the
DAG, so PETSc would never introduce Python to a DAG other than its own
build.

> This will have interesting ramifications for concretization.
> Currently when you load up a dependency spec from a `spec.yaml`, you
> get ALL of its dependencies.

The yaml file will have to specify the kind of dependency, so this is
not that different. Do these files already have a `version` field?

>                              You probably want to remove the build
> deps from the dependency's DAG and only force transitive link and run
> deps to be included in the new root's DAG, along with the root's build
> deps.  That way you can avoid the weird PETSc issue above.  If you
> include build deps in the version hash, I think you have to force
> consistency of build deps across the DAG.  That, or you have to start
> including multiple specs with the same name in a single spec DAG, but
> that complicates things like `spec['python']` --- which Python should
> it return?  I'd rather not get into that because it seems like
> needless extra complexity.
> 
> So, my take is that the build deps for package A should be documented
> in `A/.spack/spec.yaml`, but they shouldn't be included in
> `dag_hash()` or included automatically as transitive dependencies.

I think we're on the same page here too: build-only deps are invisible
except when building the package with that dependency.

> I haven't fully thought through run dependencies, but an example would
> be that `curl` is a run dependency of `git`, and decent versions of
> `tar` would be a run dependency of `spack`.  We could talk about
> whether run deps should be hashed or not... but if you don't hash them
> then we would need a better mechanism to load the right run
> dependencies in modules.  Spack doesn't currently have a way to
> determine compatibility in those cases.

These tools tend to be "just have one available" rather than any kind of
specifics. The problem is usually "it's a BSD coreutils set" where all
kinds of GNU flags and conveniences (*shakes fist at OS X's `find` not
defaulting to searching `.`*) aren't available (though portable programs
shouldn't be using them blindly anyways).

>           and I like the term "scope" for this.

Yeah, that sounds better.

> Maven's `provided` scope is an interesting one, as it corresponds to
> what people are doing when they explicitly leave a dep out of
> `package.py` to pull in the system version of it.  Personally, I think
> we don't need this, and it can be taken care of by the externals
> support in #120 and #309.  In other words, I think `provided` is a
> property of the dependency and not of the relationship between the
> dependent and the dependency.

+1

> That said, to make externals more usable, it would be nice to
> generalize detection of external build and run dependencies from the
> current way compilers are detected.  #309 adds support for multiple
> compiler detection strategies -- perhaps we could add system package
> detection strategies in a similar way.

Packages could have methods which detect if a version of the package is
available already. I will say that, from experience, these things are a
PITA to maintain across all kinds of disparate environments.

mathstuf commented:
Down to 9 test failures…

mathstuf commented:
Down to 8 errors now. It seems to be something about concretizing the specs where the concrete version of the spec is a duplicate dependency when added. I can't seem to find where in the diff something changed related to this; specs are copied where they used to be copied, deleted where they used to be deleted, and so on. Where else are such bits manipulated?

mathstuf commented:
Tests all pass locally. Any comments on implementation before I go back and edit commit messages? Any TODO items which can be removed?

tgamblin commented:
@mathstuf: I'm going to try to get around to reviewing this before the telcon tomorrow, but if I don't then I'll definitely get to it before Friday.

mathstuf commented:
Successfully built qhull after changing its dependency type to `DEP_BUILD`.

mathstuf commented:
Updated with some bug fixes (now putting them into separate commits, but can squash them back if wanted). I've also gone through and update dependencies to be more accurate (Python module dependencies are build and run deps, but not install deps; autotools and cmake are build deps; Python modules from C++ programs are run deps (ParaView)).

mathstuf commented:
@tgamblin Ping? :)

tgamblin commented:
Reviewing it this afternoon! :smile:

tgamblin commented:
ok so by this afternoon I meant late tonight, but here goes.

tgamblin commented on lib/spack/spack/build_environment.py:
```diff
@@ -196,11 +196,9 @@ def set_module_variables_for_package(pkg, m):
-    m.cmake = which("cmake")
+    def _cmake(*args, **kwargs):
+        which('cmake')(*args, **kwargs)
+    m.cmake = _cmake
```
Could also just do `m.cmake = Executable('cmake')` here.

tgamblin commented on lib/spack/spack/build_environment.py:
```diff
@@ -136,7 +136,7 @@ def add_env_path(path):
     # Prefixes of all of the package's dependencies go in
     # SPACK_DEPENDENCIES
-    dep_prefixes = [d.prefix for d in pkg.spec.traverse(root=False)]
+    dep_prefixes = [d.prefix for d in pkg.spec.traverse(root=False, deptype=spack.DEP_BUILD)]
```
`SPACK_DEPENDENCIES` is actually used in `cc` to set RPATHs so i think you want to change this to `spack.DEP_INSTALL`

tgamblin commented on lib/spack/spack/build_environment.py:
```diff
@@ -242,9 +240,10 @@ def set_module_variables_for_package(pkg, m):
     """Get a list of all the rpaths for a package."""
     rpaths = [pkg.prefix.lib, pkg.prefix.lib64]
-    rpaths.extend(d.prefix.lib for d in pkg.spec.dependencies.values()
+    deps = pkg.spec.dependencies(deptype=spack.DEP_INSTALL)
```
this makes sense to me but per the comment above it won't match what is in the `cc` wrapper

tgamblin commented on lib/spack/spack/cmd/test-install.py:
```diff
@@ -114,8 +114,12 @@ def fetch_log(path):
-    return set(childSpec for childSpec in spec.dependencies.itervalues() if not
-        spack.repo.get(childSpec).installed)
+    def get_deps(deptype):
+        return set(childSpec for childSpec in spec.dependencies(deptype)
```
Seems like dependencies() should take a `deptype_query` like `traverse()`

mathstuf replied:
`dependencies` is a shallow query, so there's no `deptype_query` to make sense.

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -161,7 +161,8 @@ def concretize_compiler(self, spec):
         try:
-            nearest = next(p for p in spec.traverse(direction='parents')
+            nearest = next(p for p in spec.traverse(direction='parents',
+                                                    deptype=(spack.DEP_BUILD, spack.DEP_INSTALL))
```
what's the difference between `deptype` and `deptype_query`?  shouldn't external code use only one of these?

mathstuf replied:
`deptype` => what dependencies of the root do I look at?
`deptype_query` => what dependencies of those found do I look at?

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -393,7 +393,7 @@ def _add(self, spec, path, directory_layout=None):
         else:
             self._data[key] = InstallRecord(spec, path, True)
-            for dep in spec.dependencies.values():
+            for dep in spec.dependencies(spack.DEP_INSTALL):
```
The ref count is intended to keep around the record in the DB until nothing refers to the missing spec.  If you set the ref count only based on *install* dependencies, then you can still run a `spack find -d` and get errors, because remaining `spec.yaml` files may refer to a build or run dependencies that have since been uninstalled. You want to keep the record in the DB to handle cases like this -- the DB is supposed to tell us not only what is installed but also what is missing, and it keeps refs around until nothing could refer to the spec anymore.

Aside: Even if the ref count were just used to track what should be uninstalled, this would be close, but I think you would want to track both install and run dependencies here, as you don't want to allow something to be uninstalled if it's still a run dependency of something else.

mathstuf replied:
Yeah, adding `DEP_RUN` should be done here. Still not convinced that build dependencies need to be kept around, but I'll do so and add a comment.

mathstuf replied:
One reason to drop pure-build deps is that there's no reason to ship autoconf, CMake, py-setuptools, etc. with your final application's RPM.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -74,6 +75,15 @@ class OpenMpi(Package):
 
 
+#
+# Types of dependencies.
```
I think these are a better fit in `spack.spec`, since they're fundamental to how you consider the spec DAG (and they're used in query functions)

mathstuf replied:
Will move.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -182,10 +192,28 @@ def _depends_on(pkg, spec, when=None):
+        # The default deptype is build and install because the common case is
+        # to build against a library which then turns into a runtime dependency
+        # due to the linker.
+        # XXX(deptype): Add DEP_RUN to this? It's an uncommon dependency type,
```
I like adding `DEP_RUN`, both for backward compatibility and because a lot of libraries come with things that you may want to run with them (`libdwarf` comes with `dwarfdump`, etc. etc.).  We'll eventually use `DEP_RUN` to set up the module environment for each package, and in that case you probably do want the binaries (at least from `bin`) in the path.  I also think it's good for backward compatibility.

Can you change `DEP_INSTALL` to `DEP_LINK` or `DEP_LIB`?   I think that is closer to the intended purpose of the type.  `INSTALL` is a little ambiguous, since technically run dependencies need to be installed to run, and build dependencies need to be installed at build time.

mathstuf replied:
The reason I kept "link" out of it is because of things like Java which don't *link*, but do need the jarpath set up similarly to find dependent `.jar` files (I think, not *that* familiar with Java). But I guess such things are similar enough to "link" that Java (packaging) folks would understand it.

As for `DEP_RUN`, the backwards compat is nice, but is /so/ rare that having `depends_on` mention it to *not* add a run dependency is noisy. Finding things which actually care about things like `dwarfdump` being in the `PATH` is easy and makes the `depends_on` lines a lot more informative.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -182,10 +192,28 @@ def _depends_on(pkg, spec, when=None):
     if pkg.name == dep_spec.name:
         raise CircularReferenceError('depends_on', pkg.name)
 
+    pkg_deptypes = pkg._deptypes.setdefault(dep_spec.name, set())
```
I like `setdefault` and all but I hate that it's not short-circuiting.  `set` construction is fast, I suppose, but the biggest current bottleneck in Spack is reading all the `package.py` files so every little bit helps...

mathstuf replied:
The `set` is not what I'd be worried about. Instead, it's the two dictionary lookups:

```python
if 'foo' not in d:
    d['foo'] = local = Default()
else:
    local = d['foo']
```

It would be nice if dictionaries had `setdefault_with` which took a 0-ary function, but alas…

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -182,10 +192,28 @@ def _depends_on(pkg, spec, when=None):
         raise CircularReferenceError('depends_on', pkg.name)
 
+    pkg_deptypes = pkg._deptypes.setdefault(dep_spec.name, set())
+    map(pkg_deptypes.add, deptypes)
```
why use `map` when you're discarding the result?

mathstuf replied:
The result is a list of `None`. The idea is to add all of the provided deptypes to the deptypes which might be provided elsewhere. Example:

```python
depends_on('python', deptypes=DEP_BUILD)
depends_on('python', when='+python_bindings', deptypes=DEP_INSTALL)
```

mathstuf replied:
Also, I've done a decent amount of Haskell, so the higher-order functions are preferred when I write such things. I'll unroll it.

tgamblin commented on lib/spack/spack/graph.py:
```diff
@@ -80,12 +80,14 @@ def topological_sort(spec, **kwargs):
     """
     reverse = kwargs.get('reverse', False)
+    # XXX(deptype): iterate over a certain kind of dependency. Maybe color
+    #               edges based on the type of dependency?
```
This sounds cool.  But don't worry too much about `spack graph` for now -- we can always make it nicer later.

mathstuf replied:
Yeah, any of these `XXX(deptype)` bits can be turned into `TODO` before merge.

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -525,6 +525,7 @@ def extendee_spec(self):
         name = next(iter(self.extendees))
 
         # If the extendee is in the spec's deps already, return that.
+        # XXX(deptype): DEP_?
```
probably only want to search link and run deps here.  I don't think you'd ever extend something that was only a build dep -- in fact, I think it should be an error to have a build-only extension relationship.  The extension concept assumes the extendee will stick around for the lifetime of the extension -- the extension has to be linked/unlinked from the directory of the extendee.

mathstuf replied:
Done.

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1063,7 +1064,7 @@ def do_activate(self, force=False):
         # Activate any package dependencies that are also extensions.
         if not force:
-            for spec in self.spec.traverse(root=False):
+            for spec in self.spec.traverse(root=False, deptype=spack.DEP_RUN):
```
this dep type should match the `xxx(deptype) DEP_?` comment above -- it's either link, run, or both.  I think you want both, as some packages need libs from the extendee as well as things like python files.

mathstuf replied:
Then they should declare a run dependency as well.

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1110,7 +1111,7 @@ def do_deactivate(self, **kwargs):
                 if aspec == self.spec:
                     continue
-                for dep in aspec.traverse():
+                for dep in aspec.traverse(deptype=(spack.DEP_INSTALL, spack.DEP_RUN)):
```
shouldn't this match activate?

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1199,9 +1200,11 @@ def fetch_remote_versions(self):
         """Get the rpath this package links with, as a list of paths."""
         rpaths = [self.prefix.lib, self.prefix.lib64]
-        rpaths.extend(d.prefix.lib for d in self.spec.traverse(root=False)
+        # XXX(deptype): Use pkg.spec.dependencies()?
```
I think this is a good idea and it keeps the `RPATH` length shorter.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -291,6 +291,27 @@ def __repr__(self):
 
 
 @key_ordering
+class DependencySpec(object):
```
I lik this -- it's basically an edge in the DAG.  Makes sense.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -426,7 +447,56 @@ def __init__(self, spec_like, *dep_like, **kwargs):
+            self._add_dependency(spec, (spack.DEP_BUILD, spack.DEP_INSTALL))
+
+
+    def get_dependency(self, name):
```
why not call this `get_depspec` and return the dep spec?  That probably scales better to any future edge properties we want to add.

The unpacking is nice now, and you could preserve it by making `DependencySpec` have an `__iter__` or `__getitem__` method, but using unpacking won't work if DependencySpec grows more things...

mathstuf replied:
Changed it to return the `DependencySpec`. I think this might be a remnant from before I added the class.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -426,7 +447,56 @@ def __init__(self, spec_like, *dep_like, **kwargs):
+    def _deptype_norm(self, deptype):
+        if deptype is None:
+            return set(spack.DEP_ALL)
+        # Force deptype to be a tuple so that we can do set intersections.
```
isn't it actually forcing it to be a set?

mathstuf replied:
Err…yeah.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -426,7 +447,56 @@ def __init__(self, spec_like, *dep_like, **kwargs):
+
+        return [dep.spec
+                for dep in where.values()
+                if deptype and deptype.intersection(dep.deptypes)]
```
Given that there are likely to only be a handful of dep types, I suspect it is way faster to use this construction:

```python
    any(d in deptypes for d in dep.deptypes)
```

Or basically anything that doesn't require building sets.  There are going to be ~3ish compares for each deptype list... which takes way less time than constructing sets for intersection.  I worry about this one b/c we've had issues with lots of object creation in concretization before -- if there end up being a lot of these types of queries and such, it could slow the process down.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -426,7 +447,56 @@ def __init__(self, spec_like, *dep_like, **kwargs):
         for dep in dep_like:
             spec = dep if isinstance(dep, Spec) else Spec(dep)
-            self._add_dependency(spec)
+            self._add_dependency(spec, (spack.DEP_BUILD, spack.DEP_INSTALL))
```
Rather than assuming defaults here (and writing the default in yet another place) I would leave this empty to start with.

Rationale: When you read in a spec from a literal, you've got no way of knowing what dep types there are.  For all the other fields in `Spec`, the convention is that we leave them empty or `None` until either `concretize()` or `normalize()` fills them in.  Here, I would say to fill this in based on `normalize()`, because `normalize()` is supposed to merge information (like constraints) from `package.py` files with what is in the user-provided `Spec`  Since the current `package.py` tells you what the deptypes are, `normalize` seems like the right place to assign this.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -683,7 +778,7 @@ def to_node_dict(self):
     def to_yaml(self, stream=None):
         node_list = []
-        for s in self.traverse(order='pre'):
+        for s in self.traverse(order='pre', deptype=(spack.DEP_INSTALL, spack.DEP_RUN)):
```
I think we still want to write out the build deps as part of the provenance for each build.  So this ought to generate YAML that includes them but doesn't hash them.  And they should be marked as build deps...

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -709,6 +804,11 @@ def from_node_dict(node):
         for name, enabled in node['variants'].items():
             spec.variants[name] = VariantSpec(name, enabled)
 
+        # XXX(deptypes): why are dependencies not meant to be read here?
```
`from_node_dict` is used by `from_yaml`, which handles the deps -- it's just for one record.  Do you want to read them here for some reason?

mathstuf replied:
Not necessarily, it was just an oddity I noticed when adding the deptypes. This looked like a place to add code to.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -887,6 +989,13 @@ def concretized(self):
+        return flat_deps
+
+
+    def flat_dependencies_with_deptype(self, **kwargs):
```
I'm a little worried about all the `_with_deptype` suffixed things, but I don't have an obviously better idea off the top of my head.

mathstuf replied:
How about a `_full` suffix? `_depspec`?

Not sure about that case, would have to test. It might prefer only the first deptype set it comes across (cf. `_dup`).

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1750,7 +1872,7 @@ def do_parse(self):
                         self.last_token_error("Dependency has no package")
                     self.expect(ID)
-                    specs[-1]._add_dependency(self.spec())
+                    specs[-1]._add_dependency(self.spec(), (spack.DEP_BUILD, spack.DEP_INSTALL))
```
See my comment above on pre-setting the default deptype -- I think this should remain unassigned until `normalize` is called.

tgamblin commented:
Ok I think I mostly got through this, except the packages.  High level, this looks really good.  This was a lot of work -- Thanks!  

Summary of stuff that still needs to be addressed:

- [ ] There are still some inconsistencies in how the deptypes are used across the different files (activate/deactive, for instance).  Might be nice to store some of those things that need to be written twice in better-named vars, e.g. `activated_deptypes = (DEP_INSTALL, DEP_RUN)`, then use `activated_deptypes` in both activate and deactivate.
- [] Looks like the database is only refcounting install deps.  It should refcount all of them as the DB tracks installed and still-referenced packages of all types.  Need this to report missing deps in `spack find`.
- [] `spec.yaml` should output something about build deps, and they should get included in the extended provenance (this may have been merged after your current base).  I think they build deps shouldn't be hashed right now, but we still want to know about them.
- [] Specs shouldn't be set up with default deptypes -- let `normalize` do that and leave the deptype undefined to start with. deptypes should come from `package.py` files.
- [] There are probably merge issues for you with `normalize` and `concretize`, as of #549.
- [] It would be nice to get rid of `spack.DEP_BUILD`, etc. and replace with just `build` At least in the package files, that would be nicer and more like many other simple python APIs.
- [] some stuff uses `deptype`, other stuff uses `deptype_query`... could that just be `deptype`?
- [] not sure about all the `_with_deptype` suffixes -- not sure how exactly to get rid of them either, but I need to think some more about that.

I think that is all for now.  I'll probably be looking over this more tomorrow and we can talk more Friday.  Thanks again!

mathstuf commented:
So there are now docs and tests (which actually found a bug!). External pkg tests are still failing, but I'm looking into that one now.

Also changed: `DEP_` is gone in favor of strings in most places and `alldeps` and `pydeps` for common combinations.

mathstuf commented:
Oh, `install` is now `link` too.

mathstuf commented:
And fixed the external packages. Missed some uses of the old `spec.dependencies` map that needed to be fixed up.

mathstuf commented:
Rebased again…but I see there are still more conflicts. Back to the drawing board :) .

mathstuf commented on var/spack/repos/builtin/packages/julia/package.py:
```diff
@@ -20,11 +20,11 @@ class Julia(Package):
+    depends_on("cmake @2.8:", deptypes=('build', 'run'))
+    depends_on("git", deptypes=('build', 'run'))
+    depends_on("openssl", deptypes=('build', 'run'))
+    depends_on("python @2.7:2.999", deptypes=('build', 'run'))
```
@eschnett

eschnett replied:
This looks good.

There are two more dependency declarations further down; these are purely run-time dependencies.

mathstuf replied:
How are they used? Pure runtime deps on either of those without linking them is, to me, odd, but I am also unfamiliar with Julia. Are they dependencies that Julia guarantees will be there for package installation?

eschnett replied:
The package manager links against them. However, from Spack's point of view, the package manager runs at run-time.

Yes, I added these so that the package manager can find them. For HDF5, the package manager could also install its own version of HDF5 (but why should it). For MPI, integrating with a system MPI seems prudent.

The latter two (HDF5 and MPI) should probably be variants, enabled by default. I can make this change later.

mathstuf commented:
@citibeth How are the docs here?

citibeth commented:
Thanks for sharing.  I like this feature, and the documentation is clear.
I would suggest the following changes:

1. Take more time to explain the "multi" dependency tyeps at the end
(alldeps, pydeps, etc).

2. Consider getting rid of abbreviations other than "alldeps", which I
would change to just "all."  I think they are hard to remember and
therefore make packages less clear.

3. Specify deptypes all in one string.  This will make life easier for the
Package author, and lessen the desire for multi-type abbreviations.  For
example:

    depends_on("libelf", deptypes=("build,link"))


There's been a lot of talk about module files recently, and this brings up
for me:

 1. I wonder if Spack should be loading a module file in response to a
build or link dependency --- rather than trying to figure out on its own
how it should change env vars.  As the modules become more flexible, this
could become more of an issue (meaning... the RIGHT way to change env vars
for a particular package may not be the way that's hardcoded into this part
of Spack).

2. Specification of a run dependency would be a great hint to any recursive
module system to recursively include that module as well (PR #721, #744).


On Thu, Apr 21, 2016 at 4:59 PM, Ben Boeckel <notifications@github.com>
wrote:

> @citibeth <https://github.com/citibeth> How are the docs here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/LLNL/spack/pull/378#issuecomment-213112686>
>

mathstuf commented:
The problem with `all` is that it is a Python builtin function and we have a pattern of `from spack import *`, so it would clobber.

As for the multi-type abbreviation, there shouldn't be many multi-type dependencies since the default has "build" and "link" in it (by far the most common deptype). Python is next with "build" and "run" which is why it has a convenience variable. Other than that, I think there are a total of 6 multi-deptype packages (all "build" and "run", but not Python dependencies, so I didn't use `pydeps`).

I agree about the module bit, but this will only help that if/when it comes about. It would make things like Ruby support easier (`RUBYLIB` needs populated) without changing spack.

mathstuf commented:
> I agree about the module bit, but this will only help that if/when it comes about.

Meaning: I don't think that change belongs in this branch :) .

mathstuf commented:
@citibeth docs updated.

mathstuf commented:
@tgamblin Do you have time to look at this for a merge?

tgamblin commented:
@mathstuf: planning for tomorrow afternoon -- I really want this in too.

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -102,6 +102,7 @@ def cmp_externals(a, b):
         return usable
 
 
+    # XXX(deptypes): Look here.
```
is this supposed to get fixed?  not sure what it's for

tgamblin commented on lib/spack/spack/cmd/package-list.py:
```diff
@@ -80,11 +80,18 @@ def print_rst_package_list():
+        for name, deptype in (
+                ('Build', 'build'),
+                ('Link', 'link'),
+                ('Run', 'run')):
```
why not just:

```python
name = deptype.capitalize()
```

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -269,20 +270,22 @@ def concretize_compiler(self, spec):
         return True  # things changed.
 
 
+# XXX(deptype): seems to be improper.
```
this one too -- what's improper?

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -319,7 +324,10 @@ def _check_ref_counts(self):
-            for dep in rec.spec.dependencies.values():
+            # XXX(deptype): This checks all dependencies, but build
+            #               dependencies might be able to be dropped in the
+            #               future.
```
Is this consistent with the ref counting?  It looks like you only increment for link here, but this checks for all.

mathstuf replied:
Dependency queries get all by default. Only traversal defaults to `('build', 'link')`.

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -393,7 +401,7 @@ def _add(self, spec, path, directory_layout=None):
         else:
             self._data[key] = InstallRecord(spec, path, True)
-            for dep in spec.dependencies.values():
+            for dep in spec.dependencies('link'):
```
Should this be *only* link?  I can really only see pure build deps getting dropped here, as they are actually no longer needed by dependents after the build step.  `link` *and* `run`, however, will still be needed by installed dependents, so shouldn't this increment for both `link` and `run` deptypes? If you take the run dep away, Spack can't guarantee that the thing will actually run.

mathstuf replied:
Oops, indeed.

tgamblin commented on lib/spack/spack/test/spec_dag.py:
```diff
@@ -437,3 +437,73 @@ def test_copy_concretized(self):
+    """
+    Here is the graph with deptypes labeled (assume all packages have a 'dt'
+    prefix). Arrows are marked with the deptypes ('b' for 'build', 'l' for
+    'link', 'r' for 'run').
```
nice!

tgamblin commented:
@mathstuf: this looks pretty good to me.  I like `pydeps` and `alldeps` mostly, but would `pydeps` make more sense as `nolink` or something?  Seems more general.

* Can you take a look at the comments i added about ref counts? I think refcounts should include run as well as link deps.

Aside from that I am not sure anything else is needed.  It seems like a few shortcuts might be nice, like `depends_build` for build deps (more readable than `deptypes=('build')`) but that could be added later.

mathstuf commented:
Yeah, renaming `pydeps` to `nolink` is better; with Lua here now, it does the same thing. Probably `R` as well, but I'm less sure there.

mathstuf commented:
OK, so I have the branch now reordered and squashed down instead of being lots of little "fixup" bits too. I also renamed `pydeps` to `nolink` and used it in more places.

mathstuf commented:
Rebased on top of the latest `develop`. I've fixed the `flake8` issues I introduced in this branch, but left anything in files touched by this branch alone because that is a *large* change (looks to be larger than the branch itself) and I'd like to keep the noise to a minimum here.

mathstuf commented:
Rebased for the architecture merge.

mathstuf commented:
Needs to update the code from #721 for API changes.

davydden commented:
maybe build dependencies should be separated from the main DAG (and build separately) and here's why:

Real-life example DAG  
```

dealii -> cmake [build] -> openssl -> zlib
   \
    \-> zlib [runtime]
```
I could argue that once there is cmake of recent enough version, i don't care which packages it used in its build. All I care is that I can now configure `dealii`. On another hand, `dealii` links against `zlib` and this is a runtime dependency which i would care about.

I think it makes sense to split this into two DAG's:
```
dealii -> cmake [build]
   \
    \-> zlib [runtime]
```
and
```
cmake [build] -> openssl -> zlib
```

Why would this be helpful? 

(i) Say i want to experiment with different compilers
`spack install dealii%clang` and `spack install dealii%gcc`. I would only care about those compiler being used consistently in runtime dependencies and the only thing i need from build time dependencies (here `cmake`) is that they are available. If the split into two DAG's is not done, every time I change compiler I would need to rebuild cmake. 

(ii) Another benefit of splitting DAGs is that this would allow `dealii%gcc^zlib%gcc^cmake%[whatever]`. In other words, it would drop some, in my oppinion, unnecessary constraints and would make packages concentrate on those part of DAGs which matter -- run-time dependencies.


Of course, i can't tell if this applies to all build dependencies, but at least for `cmake` to me this makes sense.

davydden commented:
I just noticed that @mathstuf already outlined what appears to be the same idea in https://github.com/LLNL/spack/issues/839 :smile:

mathstuf commented:
@tgamblin Ping? :)

tgamblin commented:
@mathstuf: I want to merge this with one change -- can we change `deptypes=` to `type=` in the `depends_on` directive?  I think these are more readable:

    depends_on('foo', type=('build', 'link'))
    depends_on('foo', type='link')
    depends_on('foo', type='run')

Than these:

    depends_on('foo', deptypes=('build', 'link'))
    depends_on('foo', deptypes='link')
    depends_on('foo', deptypes='run')

Also the `dep` seems redundant here b/c it's already in `depends_on`.  Any objections?  I think the rest of it looks good!

tgamblin commented:
@mathstuf: also I can make the change if you want.

davydden commented on lib/spack/spack/directives.py:
```diff
@@ -180,24 +180,43 @@ def _depends_on(pkg, spec, when=None):
+    if deptypes is None:
+        # The default deptype is build and link because the common case is to
+        # build against a library which then turns into a runtime dependency
+        # due to the linker.
```
looks like an outdated comment. From the description of `run` and `link`, i got an impression that `run` is primarily for extensions of python and alike.

i suppose `mpi` would be `build`+`link`+`run`?

tgamblin replied:
Probably not, actually.  The package *itself* does not call `mpirun` at runtime.  The user calls that *before* the app runs.  So, MPI is just a build and link dependency of the application.  A run dependency would be something like `curl` for `git`.

davydden replied:
`curl` and `git` are probably `build` dependencies?

I think for distant observer like me it is not clear what `run` dependencies are in the context of package manager like Spack? You don't normally use/run application using Spack? Maybe this could be clarified in the description of `run` at the top.

tgamblin replied:
An easier way to think about this is whether, when `A` -> `B`, `B` needs to be in the `PATH` when you `module load A` in order for `A` to work.

davydden replied:
If I am not mistaken `run` is thereby used exclusively in `spack load` and should not affect environment of `spack install` at all? I think example you mentioned will be helpful for package developers to understand the difference. So i would mention it in the description.

tgamblin replied:
Yep.

To understand this you have to distinguish runtime from build time.  `run` dependencies are for when a user runs the installed package.  `build` dependencies are for when *spack* builds the package.  Those are two different environments.  This is the same reason why things like `Package.setup_environment()` distinguish between `spack_env` and `run_env`.

alalazo commented:
@tgamblin 
> I want to merge this with one change -- can we change deptypes= to type= in the depends_on directive? 

In that way you'll mask builtin `type`...

tgamblin commented:
>> I want to merge this with one change -- can we change deptypes= to type= in the depends_on directive?

> In that way you'll mask builtin type...

Only within the `depends_on` directive, where it's in scope, and if we have to we can use `**kwargs` to avoid that.  In a package, `kwargs` don't pollute the surrounding scope.

mathstuf commented:
(Still building Qt.)

mathstuf commented:
Qt built successfully.
